### PR TITLE
Update libcoap to 4.3.1 and fix some issues with the Build

### PR DESCRIPTION
**Pull request by pulsastrix**
This PR fixes some build issues that arose because libcoap didn't necessarily always use the same crypto library version when compiling as the rust linker.

Additionally, this PR updates libcoap to the stable release 4.3.1.

### CODE REVIEW

File: libcoap-sys/build.rs [hunk 1]
```diff
@@ -114,40 +114,55 @@ fn main() {
...
+                    // Set include path according to the path provided by openssl-sys (relevant if
+                    // openssl-sys is vendored)
+                    if let Some(openssl_include) = env::var_os("DEP_OPENSSL_INCLUDE") {
+                        build_config.env("OpenSSL_CFLAGS", format!("-I{}", openssl_include.to_str().unwrap()));
+                    }
+                },
+                DtlsBackend::MbedTls => {
+                    // Set include path according to the path provided by mbedtls-sys (relevant if
+                    // mbedtls-sys is vendored).
+                    // libcoap doesn't support overriding the MbedTLS CFLAGS, but doesnt set those
```
Comment (falko17) [inline libcoap-sys/build.rs:155]:
```suggestion
                    // libcoap doesn't support overriding the MbedTLS CFLAGS, but doesn't set those
```

File: libcoap-sys/build.rs [hunk 2]
```diff
@@ -114,40 +114,55 @@ fn main() {
...
+                    // either, so we just set CFLAGS and hope they propagate.
+                    if let Some(mbedtls_include) = env::var_os("DEP_MBEDTLS_INCLUDE") {
+                        build_config.env("CFLAGS", format!("-I{}", mbedtls_include.to_str().unwrap()));
+                    }
+                },
+                DtlsBackend::GnuTls => {
+                    // GnuTLS doesn't provide include directory metadata, but doesn't support
+                    // vendoring the library either. Instead, it just uses the GnuTLS found via
+                    // pkg-config, which is the same which is already found by libcoap by default.
```
Comment (falko17) [inline libcoap-sys/build.rs:164]: Sentence is a little hard to parse for me, maybe like this?
```suggestion
                    // pkg-config, which is already found by libcoap by default.
```

**Review by falko17 — APPROVED** [review]

File: libcoap-sys/build.rs [hunk 3]
```diff
@@ -141,27 +141,57 @@ fn main() {
...
                     if let Some(mbedtls_include) = env::var_os("DEP_MBEDTLS_INCLUDE") {
                         build_config.env("CFLAGS", format!("-I{}", mbedtls_include.to_str().unwrap()));
+                        build_config.env(
+                            "PKG_CONFIG_PATH",
+                            Path::new(mbedtls_include.as_os_str())
+                                .parent()
+                                .unwrap()
+                                .join("lib")
+                                .join("pkgconfig")
+                                .into_os_string(),
```
Comment (falko17) [inline libcoap-sys/build.rs:187]: Maybe this is overeager generalization, but since this chain occurs in the code three times, it could be moved into a function.

**Review by falko17 — APPROVED** [review]

File: libcoap-sys/build.rs (hunk 3 above)
Comment (pulsastrix) [reply libcoap-sys/build.rs:187]: I think that generalizing this would make more work than just keeping it as is.
